Storing embedded keyword sourcename in output.xml as sourcename property

MaciejWiczk opened the pull request:
PR for change #3750

MaciejWiczk commented:
Here are proposed changed, that will implement additional property logged in output.xml for embedded keywords only. 
Current output.xml will look like this:
```
<kw name="Add ${a} To ${b}">
<msg timestamp="20201104 11:01:46.362" level="TRACE">Arguments: [ ${n}='1' | ${m}='2' ]</msg>
```
And new output.xml will look like this:
```
<kw name="Add ${a} To ${b}" definition="Add ${n} To ${m}">
<msg timestamp="20201104 11:03:52.179" level="TRACE">Arguments: [ ${n}='1' | ${m}='2' ]</msg>
```
Currently one acceptance test fails, but it's also failing in master branch, and I do not see a way to fix it, and a change that introduced it.

MaciejWiczk commented:
@pekkaklarck, could I ask for review? I know that there is a need to change xml schema, and I can add additional changes in #3727 since they are connected or any other recommended way.

pekkaklarck commented:
Sorry for late review due to having other bigger tasks like IF/ELSE under development. Quick comments now:

- Adding the information about the original keyword name to output.xml sounds good to me in general.
- I'm not sure is `definition` the best name for it but I cannot immediately come up with anything much better. 
- I think this information should be added only to keywords in the result model i.e. `robot.result.model`, not to "base model" in `robot.model` where it would be got also by the execution side keywords. Could you test would that work as well?
- You are right that the schema needs to be updated as well but that's a small task.

pekkaklarck commented:
We can continue discussion about this on the #devel channel on our Slack as well.

MaciejWiczk commented:
@pekkaklarck, no worries, IF/ELSE is way hotter topic, I cannot wait for as well :)
- great! It will be only working for embedded keywords, both user defined or from libraries
- I believe I added it everywhere, because the information about `definition` was only available there, but I will check again
- nam is also not exactly perfect for me, bo other like `base_name` `orig_name` are even worse ;)

pekkaklarck commented on src/robot/running/model.py:
```diff
@@ -53,14 +53,15 @@ class Keyword(model.Keyword):
-                               parent)
+                               parent, sourcename)
         self.lineno = lineno
+        self.sourcename = sourcename
```
`sourcename` shouldn't be needed here at all.

pekkaklarck commented on src/robot/running/userkeywordrunner.py:
```diff
@@ -34,6 +34,7 @@ class UserKeywordRunner(object):
     def __init__(self, handler, name=None):
         self._handler = handler
         self.name = name or handler.name
+        self.sourcename = handler.name
```
There shouldn't be need to assign `self.sourcename` when we can access it via `self._handler`. Exactly like with library keywords, it ought to be possible to enhance only the `EmbeddedArgumentsRunner`.

pekkaklarck commented on src/robot/variables/tablesetter.py:
```diff
@@ -71,7 +71,7 @@ def resolve(self, variables):
     def _avoid_recursion(self):
         if self._resolving:
-            raise DataError('Recursive variable definition.')
+            raise DataError('Recursive variable sourcename.')
```
This looks like an accidental renaming.

pekkaklarck reviewed (changes requested):
Good start but changes are still needed:

- `sourcename` should only be added to result model and only be set when actually using embedded arguments.

- Tests must be added.

- The new output.xml attribute should be explained in `doc/schema/README.rst`.

MaciejWiczk commented on src/robot/output/xmllogger.py:
```diff
@@ -66,6 +66,8 @@ def start_keyword(self, kw):
         if kw.type != 'kw':
             attrs['type'] = kw.type
+        if kw.sourcename not in ('', kw.kwname):
+            attrs['sourcename'] = kw.sourcename
```
Applying the above condition would always log sourcename to output.xml, no matter if keyword is embedded or not. Perhaps I should move condition to set sourcename only if it differs form kwname/name?

pekkaklarck replied:
I think setting `sourcename` only when using embedded args is a good idea. If we only save it to output.xml in that case, we wouldn't have it with other keywords when post-processing results anyway. Looking at the code it seems this has already been done, so the condition above can also be changed to `if kw.sourcename`.

pekkaklarck commented on doc/schema/README.rst:
```diff
@@ -90,5 +90,6 @@ kw - keyword element, name is given as an attribute. Type attribute describes th
     * ``msg`` - any number of optional keyword messages
     * ``timeout`` - optional keyword timeout. Before 3.0 this was an attribute.
     * ``status`` - keyword has to have a status
+    * ``sourcename`` - name of keyword with embedded arguments, from it's definition
```
"it's" -> "its"

Could also be changed to something like

> original name of the keyword when using embedded arguments, not set otherwise. New in 4.0.

altogether.

MaciejWiczk replied:
Done

pekkaklarck commented on doc/userguide/src/ExtendingRobotFramework/CreatingTestLibraries.rst:
```diff
@@ -1667,6 +1667,10 @@ __ `Setting custom name`_
+.. note:: Storing original keyword name, as ``sourcename`` property
+          when it contains embedded arguments is a new feature in Robot Framework 4.0.
+          It allows for easier matching keywords definition with data from output.xml files.
+
```
- I don't think this should be a note because this topic hasn't been discussed otherwise. Could be just a regular paragraph explaining that the original name of the keyword is stored as `sourcename` in the output file starting from RF 4.0. Or at least the note should be worded differently like "To ease finding keyword usages from output files, the original name of the keyword is stored ....".
- This should be explained in the section discussing embedded arguments, not in the general `@keyword` decorator section. This applies also to user keywords so perhaps this should be mentioned in the section discussing using embedded args, not here in the creating libraries section.
- This is documented in the schema so I'm not 100% sure is a separate doc needed here in the UG at all. A short paragraph/note certainly wouldn't hurt.

pekkaklarck reviewed (approved):
1. Code changes look very good but one "if" condition could still be simplified. See line comments for details.
2. Great tests!
3. Documentation needs some tuning still. See line comments for details.

I hope you @MaciejWiczk could still take a look at the smallish issues mentioned in the line comments. If you can get them sorted out I can merge this possibly before beta 2 that's planned for Wednesday. If not, I can merge the PR in the current form and fix these myself after the beta.

MaciejWiczk commented:
> 1. Code changes look very good but one "if" condition could still be simplified. See line comments for details.

I don't know why I didn;t push this change, fixed now, working like a charm :)
> 2. Great tests!

Thanks! Seems like those years as a tester were not wasted :D
> 3. Documentation needs some tuning still. See line comments for details.

Yeah, I see that they are far from perfect. It is not so simple to follow the right path. Sorry bout that. 
> I hope you @MaciejWiczk could still take a look at the smallish issues mentioned in the line comments. If you can get them sorted out I can merge this possibly before beta 2 that's planned for Wednesday. If not, I can merge the PR in the current form and fix these myself after the beta.

I hope that I fixed everything that was needed in this PR, and it could make it to the beta2 :)
Thanks for great review :)

pekkaklarck commented:
Looks good! There was a conflict due to recent changes to result model but I fixed that.
